Account login

calvinlu3 opened the pull request:
This is related to https://github.com/oncokb/oncokb/issues/2943

- Can use https://github.com/calvinlu3/oncokb-keycloak to setup keycloak server with mysql and custom login theme
- Setup Google SSO
- Create keycloak custom theme that removes username/password form
- Store spring session in redis

zhx828 commented on src/main/java/org/mskcc/oncokb/transcript/service/UserService.java:
```diff
@@ -78,10 +87,125 @@ public void deleteUser(String login) {
+            user.setId((String) details.get("uid"));
+            user.setLogin((String) details.get("sub"));
+        } else {
+            user.setId((String) details.get("sub"));
```
I think sub should always be assigned to login. More to discuss.

zhx828 commented on src/main/java/org/mskcc/oncokb/transcript/service/UserService.java:
```diff
@@ -78,10 +87,125 @@ public void deleteUser(String login) {
+    @Transactional
+    public AdminUserDTO getUserFromAuthentication(AbstractAuthenticationToken authToken) {
+        Map<String, Object> attributes;
+        if (authToken instanceof OAuth2AuthenticationToken) {
```
How does the JWT Auth work now? I tried to use a self generated token, but I think the class is UsernamePasswordAuthenticationToken.

zhx828 reviewed (changes requested):
- The user id field changed to String which is a big change. I see the oauth will assign the session id to the id field. But is it necessary?
- Let's do a global search and replace for the keycloak port?

calvinlu3 commented:
- Removed jwtDecoder and CustomClaimsConverter (and other files) that are not used
- Only users that are in application db and activated will be allowed to use the site
- Fixed issue with custom JWT not working with the GET /api/account endpoint
- Changed User ID field back to Long type
- Change keycloak port to 8080 and adjusted jhipster naming

zhx828 commented on src/main/java/org/mskcc/oncokb/transcript/service/dto/UserDTO.java:
```diff
@@ -1,24 +1,71 @@
+    private String lastName;
+
+    @Email
+    @Size(min = 5, max = 254)
```
We should double check the Size limits vs liquibase definition. This email here allows 254, but liquibase has it as 191. Could you go over all limitations and see what causes the discrepancy ?

calvinlu3 replied:
This seems to be related to [issue from jhipster](https://github.com/jhipster/generator-jhipster/issues/8604). 
Since we're using utf8, then the max length for our unique email field can be 255 characters instead of 191. Should be okay to increase the liquibase constraint to 255.

zhx828 reviewed (approved):
Once you fixed all my comments, feel free to merge.

calvinlu3 commented on src/main/java/org/mskcc/oncokb/transcript/security/CustomOAuthSuccessHandler.java:
```diff
@@ -0,0 +1,50 @@
+
+        Optional<User> user = userRepository.findOneByEmailIgnoreCase(keycloakUser.getEmail());
+        // If user is in db, then let spring security handle authentication success.
+        if (user.isPresent() && user.get().isActivated()) {
```
We were going to add a ROLE_CURATOR to public, so we can update this to check for that role when we use that table.

I will make a note of this in a comment.

calvinlu3 commented on src/main/java/org/mskcc/oncokb/transcript/service/UserService.java:
```diff
@@ -78,10 +91,56 @@ public void deleteUser(String login) {
+            user.setAuthorities(getAuthoritiesFromAuthToken(authToken));
+            user.setLogin(authToken.getName());
+            user.setActivated(true);
+            return Optional.of(new UserDTO(user));
```
I should check if the user exists in db here.

calvinlu3 commented on src/main/docker/realm-config/oncokb-curation-realm.json:
```diff
@@ -0,0 +1,2213 @@
+    },
+    {
+      "id": "1acf7ad0-68cb-46a6-a3e4-8b2f2abecb85",
+      "clientId": "jhipster-control-center",
```
Do we need the [jhipster control center](https://www.jhipster.tech/jhipster-control-center/)? We're not using this feature from jhipster, so can delete this client.

There are still a couple jhipster names in this file. Will see which ones are needed, if not then they can be deleted.
